### PR TITLE
fix(#733): canonical charge representatives through the boundary (2/4)

### LINKED ISSUE #733: Bond charge from a single OUT left leg is not canonicalised (Z_n mislabelled, ProductSymmetry broken)
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

## Summary

`_group_blocks_by_bond_charge` (`src/tenax/linalg.py:95`) computes the bond charge as

```python
effective = [np.array([int(tensor.indices[i].flow) * int(key[i])]) for i in left_leg_positions]
q = int(sym.fuse_many(effective)[0])
```

When `left_leg_positions` has **one** entry, `fuse_many` receives a single array and returns it
**unreduced** — no `% n` is applied. So for a modular symmetry the bond charge is a
non-canonical representative whenever that single left leg flows OUT.

## Measurement

```
>>> ZnSymmetry(2).fuse_many([np.array([-1])])          # one array
array([-1])
>>> ZnSymmetry(2).fuse_many([np.array([-1]), np.array([0])])   # two
array([1])
```

Which propagates all the way to a public result — `tenax.linalg.svd` on a Z2 matrix,
`left_labels` one leg:

| left leg flow | bond sectors | `U` block keys |
|---|---|---|
| IN  | `[0, 1]` ✅ | `[(0,0), (1,1)]` |
| OUT | `[-1, 0]` ❌ | `[(0,0), (1,-1)]` |

Same tensor, same decomposition, different charge labels. `-1` is not a Z2 charge; the
canonical representative is `1`.

## Why it survives

`ZnSymmetry.fuse` reduces mod `n`, so `(-1 + 1) % 2 == 0` and `_validate()` passes. The
labels are self-consistent within any one decomposition, so round-trips and `todense()`
comparisons agree. What breaks is *matching against a tensor built with canonical labels*:
`contract` keeps the charge-conserving components and silently zeroes the rest rather than
raising (measured separately at 3.09 out of an 8.81-scale result on a `T1 x a` contraction,
120 of 1024 entries retained), so the failure is quiet.

`dual` is also not an involution on the non-canonical representative:
`ZnSymmetry(2).dual(-1) == 1`, and `dual(dual(-1)) == 1 != -1`. Anything that daggers such a
leg lands on different labels than it started with.

## Scope

`_group_blocks_by_bond_charge` has **7 call sites** in `linalg.py` (lines 255, 678, 938,
1166, 1343, 1552, 1943) — `svd`, `qr`, `eigh` and friends. Any of them called with a
single OUT left leg on a `Z_n`/`FermionParity` tensor gets non-canonical bond labels.
U(1) is unaffected (`dual(q) == -q` exactly).

**`ProductSymmetry` is worse than non-canonical — it is wrong.** Charges are bit-packed
(`encode = (q2 << 16) | (q1 & 0xFFFF)`), so negation is not the component-wise inverse:

```
encode(1,2) = 131073        decode(-131073) = (-1, -3)      # raw negation
dual(131073) = -65537       decode(-65537)  = (-1, -2)      # correct
fuse(q, -q) = -65536        # not the identity -> conservation genuinely violated
```

## Suggested fix

Canonicalise in `_group_blocks_by_bond_charge` — fuse against the identity, or apply the
symmetry's own reduction, so a single-leg group returns the same representative a
multi-leg one would. That is one place rather than seven, and it makes the single-leg and
multi-leg paths agree by construction.

A regression test worth having: for each of U(1), Z2, Z3 and `ProductSymmetry`, assert
`svd` produces the same bond sectors with the left leg IN and with it OUT.

## Found via

#729 (#715 Phase 3). The symmetric root-implicit module builds its cut projectors with
row and col legs flowing the *same* way, which is exactly the orientation that exposes
this, and it reproduces the library's `-1` labelling because diverging from it would make
its bonds fail to contract against the rest of tenax. That module now refuses
`ProductSymmetry` outright and pins the Z_n convention in
`test_zn_same_flow_bonds_carry_the_library_s_own_representative`, which is the test that
should fail and be updated in step if this issue is fixed.

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

**PR 2 of 4** for #734. Stack: PR1 (#736) → **PR2** → PR3 → PR4. **Base is #736** — merge that first.

**Closes #733.**

## What

- `TensorIndex.__post_init__` canonicalises sectors and **merges duplicates**. Z2 `[-1, 0, 1]` becomes `sectors=[0, 1]`, `multiplicities=[1, 2]`, `dim=3`.
- `from_charges` / `dual()` canonicalise the cached dense charges, so `charges` cannot desynchronise from `sectors`.
- `_validate` and four `linalg.py` sites go through `_net_charge`.
- `_compute_valid_blocks` **drops its `is_infinite = n_values() is None` split entirely**, replacing both branches with one closed form.

## The bug (#733)

`_group_blocks_by_bond_charge` fused a *single* flow-weighted charge, and `fuse_many` of one array returns it unreduced — so no `% n` was applied:

| symmetry | left leg IN | left leg OUT |
|---|---|---|
| Z2 | `[0, 1]` | `[-1, 0]` ❌ |
| Z3 | `[0, 1, 2]` | `[-2, -1, 0]` ❌ |
| FermionParity | `[0, 1]` | `[-1, 0]` ❌ |
| U(1) | `[0, 1]` | `[-1, 0]` ✅ canonical for U(1) |

Same tensor, same decomposition, different labels. `-1` is not a Z2 charge. Labels that disagree make `contract` silently keep the charge-matching components and zero the rest rather than raising.

## Why the branch split could go

With `flow_charge` an involution on canonical representatives (PR 1), `q_last = flow_charge(flow, fuse(target, dual(partial)))` is valid for **any** abelian group. Cross-checked against exhaustive enumeration over **1470 configurations** — 0 mismatches. The only divergences from the old code are **6 cases where the old code was wrong**: rank-1 finite-group legs with a non-identity target returned `[]`.

That also means the `ProductSymmetry(Z2, U1) → n_values() is None → U(1)-only integer algebra on packed charges` defect disappears rather than being patched.

## Scope note

Canonicalisation forced migrating `_ctm_root_implicit_sym_sectors.py` from raw `-q` keys to `flow_charge`/`dual` — the plan had scheduled that for PR 4. It builds its bond index directly from `_group_blocks_by_bond_charge`'s now-canonical `q`, so `-q` keys named charges the leg does not carry and 15 tests failed. Review verdict: **forced and minimal**.

Every changed assertion was diagnosed as a correct consequence of canonicalisation or a strengthening. Nothing was skipped, xfailed or deleted.

## Review-driven changes

- **`__post_init__` left already-canonical-but-unsorted sectors unsorted**, which silently broke `multiplicity`/`has_sector`/`sector_offset` — all three use `searchsorted`. `sectors=[1,0]` gave `multiplicity(1) == 0` and `sector_offset(0)` raised `KeyError`. The short-circuit that fixes it also pays for the constructor cost (7.7 → 5.6 µs).
- **`_compute_valid_blocks` had no coverage**: a full revert killed **0 of 439** tests. Now **8**.
- Canonicalising sectors opened a `sectors != charges` window in `fuse_indices`, where `ProductSymmetry` blocks on an unrepresented charge were dropped silently. Closed.
- `_validate` vectorised — **26.6×** faster than the scalar-adapter version.
- One test asserted a plan-level error: `seen["IN"] == seen["OUT"]` is **false for U(1)**, since mirroring flows dualises the bond and `-1` is a genuine U(1) charge. Replaced with canonicality, `OUT == canon(dual(IN))`, and — new — that every block key names a sector its own leg carries.

## Verification

`pytest -m core`: **1235 passed, 8 skipped, 1556 deselected**.

Closes #733. Refs #734.

### CODE REVIEW

File: src/tenax/algorithms/_tensor_utils.py [hunk 1]
```diff
@@ -151,7 +151,14 @@ def _fuse_indices_dense(
...
     sym = idx_a.symmetry
-    fused_charges = _compute_fused_charges(idx_a, idx_b, fused_flow, sym)
+    # Canonicalise before deriving sectors.  TensorIndex.__post_init__ rewrites
+    # ``sectors`` to canonical representatives (#734), so caching uncanonicalised
+    # charges would leave ``charges`` naming sectors the index does not have, and
+    # blocks on those charges are dropped silently.  U(1)/Z_n are unaffected
+    # (``_compute_fused_charges`` already reduces mod n); ProductSymmetry is not.
+    fused_charges = sym.canonicalize_charges(
+        _compute_fused_charges(idx_a, idx_b, fused_flow, sym)
+    )
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_tensor_utils.py:161]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Tracked in **#799**. The block-key half reproduces exactly as described, on `main`:

```python
z2 = ZnSymmetry(2)
i_in  = TensorIndex.from_charges(z2, [-1, 1], FlowDirection.IN,  label="a")
i_out = TensorIndex.from_charges(z2, [-1, 1], FlowDirection.OUT, label="b")
# index sectors correctly canonicalise to [1, 1]  (#733 working as intended)

SymmetricTensor({(-1, 1): jnp.ones((1,1))}, (i_in, i_out)).todense().sum()   # -> 0.0
SymmetricTensor({( 1, 1): jnp.ones((1,1))}, (i_in, i_out)).todense().sum()   # -> 4.0
```

#733 canonicalises the index; the block key keeps the caller's representative, passes `_validate` because fusion reduces mod 2, and then `todense()` finds no `-1` basis state and returns **zero with no error**. Same caller, same charge, two different answers depending on which representative they happened to write — which is the failure mode #733 set out to remove, surviving one level down.

The `_compute_fused_charges` / `split_index` half is in the same issue; both are hand-rolled charge arithmetic that should go through the group operations #734 established.
